File: pkg/reactor/mdcs_reactor/models/task.py
import enum
import datetime

from sqlalchemy import Column, ForeignKey, String, DateTime, Enum, Text
from sqlalchemy.orm import relationship

from .generic import Model


class TaskState(enum.Enum):
    PENDING = enum.auto()
    RUNNING = enum.auto()
    CANCELLED = enum.auto()
    COMPLETED = enum.auto()
    FAILED = enum.auto()


class Task(Model):
    """
    Database model for a task to run an action.
    """

    __tablename__ = 'task'

    uuid = Column(String(22), primary_key=True)
    action_uuid = Column(String(22), ForeignKey('action.uuid'), nullable=False)
    state = Column(Enum(TaskState), nullable=False, default=TaskState.PENDING)
    created = Column(DateTime, nullable=False, default=datetime.datetime.now)
    modified = Column(DateTime, nullable=False, onupdate=datetime.datetime.now)
    output = Column(Text, nullable=False, default="")

    action = relationship('Action', back_populates='tasks')

    def __repr__(self):
        return "<Task(uuid='{0}', action_uuid='{1}', state='{2}', created={3}, modified={4}, output='{5}')>".format(
            self.uuid,
            self.action_uuid,
            self.state.name,
            self.created,
            self.modified,
            self.output)

File: pkg/remote/lib/models/task.dart
import 'package:remote/models/model.dart';

/// The state of a task.
enum TaskState {
  pending,
  running,
  cancelled,
  completed,
  failed
}

/// A task to run an action.
class Task extends Model {
  String uuid;
  String actionUuid;
  TaskState state;
  DateTime created;
  DateTime modified;
  String output;

  Task({this.uuid, this.actionUuid, this.state, this.created, this.modified, this.output});

  Task.fromJSON(Map<String, dynamic> data) {
    this.uuid = data['uuid'] as String;
    this.actionUuid = data['action_uuid'] as String;
    this.created = DateTime.parse(data['created'] as String);
    this.modified = DateTime.parse(data['modified'] as String);
    this.output = data['output'] as String;

    switch (data['state']) {
      case 'pending':
        this.state = TaskState.pending;
        break;

      case 'running':
        this.state = TaskState.running;
        break;

      case 'cancelled':
        this.state = TaskState.cancelled;
        break;

      case 'completed':
        this.state = TaskState.completed;
        break;

      case 'failed':
        this.state = TaskState.failed;
        break;
    }
  }

  @override
  String get primaryKey => this.uuid;
}
